docs(website): add database migrations documentation (#1657)
aerogear/graphback: docs/databases/migrations.md

sidebar_label: GraphQL Migrations
---

Graphback uses [graphql-migrations](https://www.npmjs.com/package/graphql-migrations) to automatically create and update tables from a GraphQL schema.

> The package is able to calculate diff between your schema and your database to only executes required changes and thus keeping your database and schema in sync.

Reviewer: ```suggestion
The library compares your database schema to your GraphQL schema and executes the required changes to keep the database structure synchronised with the GraphQL schema.
```

## Compatibility

- PostgreSQL (create and update database)
- SQLite (create database only)

## Usage

> Graphback migrations operates on business models defined in your schema: These are graphql types decorated with a [`@model`](../model/datamodel#Model) annotation. 

Reviewer: ```suggestion
GraphQL Migrations operates on business models defined in your schema: These are GraphQL types decorated with a [`@model`](../model/datamodel#Model) annotation. 
```

> The package exports an entry point which you can programatically setup in your source code and have it peform the migrations. 

Reviewer: ```suggestion
The package expose an API which you can programmatically set up in your source code and have it perform the migrations. 
```

### API

The `migrateDB` method creates and updates your tables and columns to match your GraphQL schema.
All the database operations are wrapped in a single transaction, so your database will be fully rolled back to its initial state if an error occurs.

> The method takes three arguments as descripted in [migrations options](./migrations#Options) section.

Reviewer: ```suggestion
The method takes three arguments as described in [migrations options](#Options) section.
```

```ts
import { migrateDB } from 'graphql-migrations';

const schemaText = ```
"""
@model
"""
type Note {
  id: ID!
  title: String!
}
```;

const dbConfig = {
  // Knex.js based configuration
};

migrateDB(dbConfig, schemaText, {
  // Additional options
}).then((ops) => {
    console.log(ops);
});
...
```

Assuming the above code is ran agains a PostgreSQL database, the following relations will be created:
```sql
\d
              List of relations
 Schema |    Name     |   Type   |   Owner    
--------+-------------+----------+------------
 public | note        | table    | postgresql
 public | note_id_seq | sequence | postgresql
```

And the `note` table:

```sql
 \d note
                                 Table "public.note"
 Column |          Type          |                     Modifiers                     
--------+------------------------+---------------------------------------------------
 title  | character varying(255) | not null
 id     | integer                | not null default nextval('note_id_seq'::regclass)
Indexes:
    "note_pkey" PRIMARY KEY, btree (id)
```

> ## Designing your data model

Reviewer: I think a lot of this is duplicated from the Data Model page and can quickly go out of sync with each other. Though this does have only the migration specific information. Up for discussion what we do with this.

Author: I could not find a better to do this part. I have added links above and I think it is important to note that when modifying docs for a model / annotation etc one has to verify that the migration package is not broken and this includes the docs. The chance of the two docs being out of sync are there but very minimal considering that we do not go into details about what `@model` contains for example. Having said that, any suggestions welcomed about how we can re-restructure this part.

Reviewer: What I meant by "going out of sync" was that because the same content needs to be duplicated in multiple places, the likelihood of one not getting updated by accident and becoming out of sync with the other is possible.

Though you are right, I think here there needs to be duplication, especially considering this library can be used without Graphback.

You can specify a default value using the `@default` field annotation as shown below.

```graphql
type Note {
  id: ID!
  title: String!
  """
  Define a default value
  @default(value: false)
  """
  complete: Boolean
}
```

#### Custom column type

You can specify a custom type usig the `@db` field annotation as shown below:

```graphql
type Note {
  id: ID!
  """
  Define a custom json type
  @db(type: 'json')
  """
  comments: [Comment]
}

type Comment {
  text: String
  description: String
}
```

#### Changing column length in Graphback

By default string columns will be created as `varchar` with a column length of `255`. 
This can be configured to any length using the `@db` field annnotation and the `length` argument:

```graphql
type Note {
  id: ID!
  """
  Define custom column length
  @db(length: 100)
  """
  title: String!
}
```

#### Ignore a field 

Sometimes our business model can contain more fields and some of these fields may only be there for representation purposes.
We can safely ignore generating columns for these fields using the `@db(skip: true)` annotation on the corresponding field. 

```graphql
type Note {
  id: ID!
  """
  Define custom column length

>   @db(skip: true)

Reviewer: Though GraphQL Migrations supports this, Graphback does not recognise it (in that it will still create input fields.

Author: Yeap, I was only focused on the migration part. 

I guess it is worth to add a note. WDYT?

Reviewer: I think so yeah.

For example:

```graphql
"""
@model
"""
type Comment {
  id: ID!
  """
  @index(name: "title-index")
  """
  text: String
  """
  @index
  """
  description: String
}
```

The above model will result in the following indexes being created

```sql
\d comment
                                   Table "public.comment"
   Column    |          Type          |                       Modifiers                       
-------------+------------------------+-------------------------------------------------------
 text        | character varying(255) | 
 description | character varying(255) | 
 id          | integer                | not null default nextval('comment_id_seq'::regclass)
Indexes:
    "comment_pkey" PRIMARY KEY, btree (id)
    "comment_description_index" btree (description)
    "title-index" btree (text)
```

#### Unique columns

The `@unique` annotation can be used to create an unique constraint on a specific field. Graphback will generate the constraint name
using the `<tablename>_<columnName>_unique` format.

For example:

```graphql
"""
@model
"""
type Comment {
  id: ID!
  """
  @unique
  """
  text: String
}
```

> The above model will result in the following constrant being created

Reviewer: ```suggestion
The above model will result in the following constraint being created
```

```sql
\d comment
                                   Table "public.comment"
   Column    |          Type          |                       Modifiers                       
-------------+------------------------+-------------------------------------------------------
 text        | character varying(255) | 
 id          | integer                | not null default nextval('comment_id_seq'::regclass)
Indexes:
    "comment_pkey" PRIMARY KEY, btree (id)
    "comment_text_unique" UNIQUE CONSTRAINT, btree (text)
```

### Options

> The `migrateDB` takes the three arguments, listed below.

Reviewer: ```suggestion
`migrateDB` takes the three arguments, listed below.
```

```sql
\d note
              Table "public.note"
  Column   |          Type          | Modifiers 
-----------+------------------------+-----------
 reference | character varying(255) | not null
 title     | character varying(255) | not null
Indexes:
    "note_pkey" PRIMARY KEY, btree (reference)
```

### Foreign keys

Graphback migrations will automatically create and index foreign keys once it sees relationships between business models.

#### OneToOne

```graphql
"""
@model
"""
type Profile {
  id: ID!
  """
  @oneToOne
  """
  user: User!
}

"""
@model
"""
type User {
  id: ID!
  name: String
}
```

This creates a relationship via a `userId` column in the `profile` table. You can customize the key tracking the relationship with the `key` annotation:

```graphql
"""
@model
"""
type Profile {
  id: ID!
  """
  @oneToOne(key: 'user_id')
  """
  user: User!
}

"""
@model
"""
type User {
  id: ID!
  name: String
}
```

This maps `Profile.user` to `profile.user_id` in the database.

#### OneToMany

```graphql
"""
@model
"""
type Note {
  id: ID!
  title: String!
  """
  @oneToMany(field: 'note')
  """
  comments: [Comment]
}
```

> This creates a relationship between `Note.comments` and `Comment.note` and maps to `comment.noteId` in the database. If `Comment.note` does not exist it will be generated for you.

Author: Very good comment.

  @db(type: 'json')
  """
  comments: [Comment]
}

type Comment {
  text: String
  description: String
}
```

### Column length

By default string columns will be created as `varchar` with a column length of `255`. 
This can be configured to any length using the `@db` field annnotation and the `length` argument:

```graphql
type Note {
  id: ID!
  """
  Define custom column length
  @db(length: 100)
  """
  title: String!
}
```

### Ignore a field 

Sometimes our business model can contain more fields and some of these fields may only be there for representation purposes.
We can safely ignore generating columns for these fields using the `@db(skip: true)` annotation on the corresponding field. 

```graphql
type Note {
  id: ID!
  """
  Define custom column length
  @db(skip: true)
  """
  title: String!
}
```

> NOTE: This annotation is not supported by Graphback CRUD.

### Column name

When working with database migration library it is possible to change individual database columns.
Using custom column will require manual mapping in resolver or database layer. 

```graphql
type Note {
  id: ID!
  """
  @db(name: 'note_title')
  """
  title: String!
}
```

> > NOTE: This annotation is not supported by Graphback CRUD 

Author: I was thinking of caution.

  - `updateComments` (default: `false`): by default, `migrateDB` will not create or update comments on table and columns.
  - `scalarMap` (default: `undefined`): Custom Scalar mapping
  - `mapListToJson` (default: `true`): Map scalar/enum lists to json column type by default.
  - `plugins` (default: `[]`): List of [graphql-migrations plugins](https://github.com/aerogear/graphback/blob/master/packages/graphql-migrations/src/plugin/MigratePlugin.ts) which describes queries that can be executed during migrations. 
  - `debug` (default: `false`): displays debugging informations and SQL queries.
  - `removeDirectivesFromSchema`: (default: `true`): Strips all directives from schema before processing.
  - `operationFilter`: Filter out database operations that we don't want, e.g. [prevent table deletion](https://github.com/aerogear/graphback/blob/master/packages/graphql-migrations/src/plugin/MigrateOperationFilter.ts).

### Tabel and Column Comments

The `updateComments` once set to `true` will automatically create/update table and column comments. 
Table / column comments are type or field description parsed from the model schema. 

> The parsing logic takes care of stripping off all Graphback annotations leaving only business related comments.    

Author: I applied the suggestion. 

Do we have an issue for that in graphql-metadata?

Will create / update the `note` table comments as shown below: 

```sql
\d+
                                       List of relations
 Schema |    Name     |   Type   |   Owner    |    Size    |            Description             
--------+-------------+----------+------------+------------+------------------------------------
 public | note        | table    | postgresql | 8192 bytes | Note business model table comment.
 public | note_id_seq | sequence | postgresql | 8192 bytes | 
(2 rows)
```

```sql
\d+ note
                                                     Table "public.note"
 Column |          Type          |                     Modifiers                     | Storage  | Stats target | Description 
--------+------------------------+---------------------------------------------------+----------+--------------+-------------
 title  | character varying(255) | not null                                          | extended |              | Note title.
 id     | integer                | not null default nextval('note_id_seq'::regclass) | plain    |              | 
Indexes:
    "note_pkey" PRIMARY KEY, btree (id)
```
 
> NOTE: This options may not be supported by some database. 

> ### Scalar Mapping

Author: +1. This crossed over my mind too but I had no strong opinion about it. Will do it!
